### DISCUSSION TITLE
Any plans on supporting BSDs (OpenBSD/FreeBSD)

### DISCUSSION THREAD

**Discussion by tnxz** [cirruslabs/tart#397] [General]
does tart have bsds on their roadmap

**Comment by fkorotkov** [comment]
Tart is using [`VZEFIVariableStore`](https://developer.apple.com/documentation/virtualization/vzefivariablestore) for running "Linux" VMs which theoretically means you can try to run FreeBSD and/or OpenBSD already with Tart:

```bash
curl -o freebsd.iso https://download.freebsd.org/releases/arm64/aarch64/ISO-IMAGES/13.1/FreeBSD-13.1-RELEASE-arm64-aarch64-dvd1.iso
tart create --linux freebsd
tart run --disk freebsd.iso freebsd
```

Unfortunately, the commands above make the installer go in the infinite restart loop:


https://user-images.githubusercontent.com/989066/216830091-84c17678-7a7d-4bf0-b3a3-aba092875185.mov

**Reply by fkorotkov** [reply]
It seems you might need to try to prepare a special disk for EFI as described in the [docs](https://wiki.freebsd.org/UEFI). Maybe someone from FreeBSD coomunity can weight in on how to make it work. CC @emaste

**Comment by tnxz** [comment]
It would be good if someone could get it to work.
Anyhow thanks for the pointers.

**Comment by emaste** [comment]
The final message I see in the FreeBSD boot video above is `Please set hw.efifb.address and hw.efifb.stride.` which is from the loader, so it appears we are not actually getting as far as the kernel.

There is ongoing work on FreeBSD/arm64 as both native and virt guest on the MS dev platform and Apple arm64 platforms, so we should take a look at this in due course.

**Comment by poolpOrg** [comment]
Trying to boot OpenBSD but as soon as the VM starts loading the kernel, I get:

guest has stopped the virtual machine due to error: Error Domain=VZErrorDomain Code=1 "The virtual machine stopped unexpectedly." UserInfo={NSLocalizedFailure=Internal Virtualization error., NSLocalizedFailureReason=The virtual machine stopped unexpectedly.}

Considering that I have OpenBSD running as the host OS on another mac mini m1, I thought I should be able to get it to run as a tart guest but apparently not :-(

Is there a way to make tart a bit more verbose and/or dump the current state for the VM at the time of the VZErrorDomain ?

**Comment by edigaryev** [comment]
Just tried the latest FreeBSD 14.1 (`FreeBSD-14.1-RELEASE-arm64-aarch64.raw`) on macOS Sonoma 14.5 with Tart 2.12.0, doesn't seem to get past `Trying to fire up the kernel, but no device tree blob found!`:

<img width="1136" alt="Screenshot 2024-06-24 at 19 03 16" src="https://github.com/cirruslabs/tart/assets/85709/14040761-17cf-415a-b705-353b67a5b5af">

**Comment by djgoku** [comment]
![Screenshot 2024-11-17 at 16 31 29](https://github.com/user-attachments/assets/6dde1065-c792-4ea9-be7f-dec8b22cb10a)

I was able to install OpenBSD 7.6 via install76.iso (released and current), full install, and everything but -x* worked. I was able to use the system, but without networking, it is really hard to use it.

vio0 fails with inet saying `feature negotiation failed`, but somehow inet6 gets an address? I was able to start xenodm, but it was just a black screen, and I had to reboot after that.

**Reply by edigaryev** [reply]
Unfortunately, we're tied with the VirtIO network  driver that the Apple's [`Virtualization.Framework`](https://developer.apple.com/documentation/virtualization) provides, so the most likely scenario of resolving this is probably a fix for `feature negotiation failed` on the OpenBSD's side.
